Issue 1363: Initial clippy lint testing

ozghimire opened the pull request:
Initial draft for [issue 1363](https://github.com/unicode-org/icu4x/issues/1363).

**Observation** : These lints are marked as unknown lints while running `cargo build`

```
warning: unknown lint: `indexing_slicing`
  --> components/calendar/src/lib.rs:19:9
   |
19 | #[allow(indexing_slicing, expect_used, panic)]
   |         ^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unknown_lints)]` on by default

warning: unknown lint: `expect_used`
  --> components/calendar/src/lib.rs:19:27
   |
19 | #[allow(indexing_slicing, expect_used, panic)]
   |                           ^^^^^^^^^^^

warning: unknown lint: `panic`
  --> components/calendar/src/lib.rs:19:40
   |
19 | #[allow(indexing_slicing, expect_used, panic)]
   |                                        ^^^^^```
```
^^^
Solved. I was missing clippy namespace around them like `clippy::panic`

ozghimire commented:
I added few rules to components/calendar's, does this look like the right approach ?

sffc commented:
Thanks for getting started!  Initial comments:

1. Ideally any `#[allow(...)]` overrides will annotate individual lines of code, not whole modules.
2. To do this incrementally, you could start by adding `#![deny(...)]` in the *lib.rs* file of a crate of your choice, and see if you can get that to work before you do it to all the other crates.

ozghimire commented:
I was able to configure `#![deny(...)]` in lib.rs for component/calendar. The config applied for all the files inside calendar and needed multiple `#[allow()]` for it to pass `cargo clippy`. 

This strategy seems to work and will apply to all the component items.

sffc commented on components/calendar/src/iso.rs:
```diff
@@ -394,6 +394,7 @@ impl Iso {
         fixed + (date.day.0 as i32)
     }
 
+    #[allow(clippy::unwrap_used)]
```
"allow" lints should be at the line level (not the function level), and they should be accompanied by a comment explaining why it is safe to do the unwrap/index.

ozghimire replied:
These allows are needed as per the addition of deny clippy list on lib.rs . The strategies on how to handle these should change as part of that, we can try to create issues to fix these and reference on the lines as well. The reason behind that is not having full understanding of the code and it makes it harder to complete this specific issue.

ozghimire replied:
Observation : Adding these attributes are experimental on expressions
`error[E0658]: attributes on expressions are experimental`

gregtatum commented on components/calendar/src/lib.rs:
```diff
@@ -1,7 +1,7 @@
 // called LICENSE at the top level of the ICU4X source tree
 // (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
-
+#![deny(clippy::indexing_slicing, clippy::unwrap_used)]
```
Praise: This approach looks correct to me.

ozghimire commented:
@gregtatum  Updated as pre the request. And about the panic, if enabled it fails on `provider.rs line 31`. Tried adding inline allow but didn't responded to it.

gregtatum commented on components/calendar/src/japanese.rs:
```diff
@@ -158,6 +158,7 @@ const REIWA_START: EraStartDate = EraStartDate {
     day: 1,
 };
 
+#[allow(clippy::panic)] // TODO(#NNN) Clippy exceptions need docs or fixing.
```
suggestion: I think these errors are coming from the `tinystr!` macro implementation. I would suggest adding this suppression in the `utils/tiny_str` crate.

sffc replied:
+1. It looks like a large fraction of these come from the `tinystr!` macro.

gregtatum commented on components/calendar/src/provider.rs:
```diff
@@ -27,6 +27,7 @@ pub struct EraStartDate {
 }
 
+//#[allow(clippy::panic)] // TODO(#NNN) Clippy exceptions need docs or fixing.
 #[icu_provider::data_struct(JapaneseErasV1Marker = "calendar/japanese@1")]
```
My guess is that there is an `unwrap` in the proc_macro `icu_provider::data_struct`

gregtatum replied:
```diff
diff --git a/components/calendar/src/lib.rs b/components/calendar/src/lib.rs
index df3c2aa29..2ee4cbe15 100644
--- a/components/calendar/src/lib.rs
+++ b/components/calendar/src/lib.rs
@@ -6,7 +6,7 @@
     clippy::unwrap_used,
     clippy::expect_used,
     // If enabled it fails on provider.rs line 31
-    //clippy::panic
+    clippy::panic
 )]
 #![cfg_attr(not(any(test, feature = "std")), no_std)]
 
diff --git a/provider/macros/src/lib.rs b/provider/macros/src/lib.rs
index ce71709d1..caca75fb3 100644
--- a/provider/macros/src/lib.rs
+++ b/provider/macros/src/lib.rs
@@ -111,6 +111,7 @@ fn data_struct_impl(attr: AttributeArgs, item: ItemStruct) -> TokenStream2 {
                     }
                 ));
             }
+            #[allow(clippy::panic)]
             _ => {
                 panic!("Invalid attribute to #[data_struct]")
             }
```

I'm still getting the error locally, but the suppressions need to be added inside the macro. Macros expand the code in place, so clippy rule is being applied to the expanded code, but it's being reported at where the proc_macro is defined.

gregtatum reviewed (changes requested):
Marking as request changes as this is still a draft PR, and I have answered the questions.

ozghimire commented:
Observation: 
- Added  clippy rules in `utils/tiny_str crate` but still needed these rules explicitly in `japanese.rs`.
- Added clippy rules in `provider/macros/src/lib.rs` still same issue on `provider.rs` needed to add outward declaration on the top of  the file to resolve the error.

sffc commented on components/calendar/src/iso.rs:
```diff
@@ -395,6 +397,7 @@ impl Iso {
     }
 
     fn fixed_from_iso_integers(year: i32, month: i32, day: i32) -> i32 {
+        #[allow(clippy::unwrap_used)] // TODO(#NNN) Clippy exceptions need docs or fixing.
```
Suggestion: For cases like this, please take the time to investigate the safety of the unwrap. In this case, I see that this is a private function with 3 call sites. Try to structure the code such that the clippy exception is closer to the call site where it is quickly obvious that there is no risk of panic.

sffc commented:
@Manishearth Can you take a look at the issue where enabling the clippy unwrap_used lint triggers on instances of the `tinystr!()` macro and what we can do to avoid that?

Manishearth commented:
@sffc the fix upstream is for these lints to potentially ignore stuff inside `const` contexts. but i'm unsure if people would want this in clippy (worth filing an issue!)

the fix here is to add the allow attribute to the const generated by the tinystr macro. i think we should just do that

sffc commented:
> the fix here is to add the allow attribute to the const generated by the tinystr macro. i think we should just do that

What I definitely want to avoid is someone using `tinystr!()` and then being required to add a `#[allow(clippy::unwrap_used)]` to annotate it every time. Clippy annotations should be required very sparingly. If they are used everywhere, they become harder to review and they lose their potency.

> @sffc the fix upstream is for these lints to potentially ignore stuff inside const contexts. but i'm unsure if people would want this in clippy (worth filing an issue!)

If the const context is at fault, then we could wrap the contents of the macro in a panicky function, and put the suppression inside the panicky function.

Manishearth commented:
> What I definitely want to avoid is someone using `tinystr!()` and then being required to add a `#[allow(clippy::unwrap_used)]` to annotate it every time. Clippy annotations should be required very sparingly. If they are used everywhere, they become harder to review and they lose their potency.
> 


Right, I don't want that either: I'm suggesting that we add a _single_ annotation to the _body_ of `tinystr!()` :smile:

sffc commented:
> Right, I don't want that either: I'm suggesting that we add a single annotation to the body of tinystr!() smile

That's my preference as well. I was under the impression that this was already tried and it didn't work. But it does work:

https://github.com/unicode-org/icu4x/pull/1677

sffc commented:
@ozghimire Can you merge in #1677 and see if you can get rid of the extra annotations then?

ozghimire commented:
@sffc  Merged and pushed up with changes. Provider macro is interesting, even after inline allow for panic clippy it still needed file wide allow on `canander/src/provider.rs`

sffc commented:
@ozghimire There seems to be an issue with your PR; the diff when merged into main should be only your changes, not changes from other people's PRs. This is most likely something you can fix in `git`. If you need help with `git`, there are many resources you can find with a quick Google search for something like "git tutorials".

ozghimire commented:
@sffc @gregtatum This changes are updated in the entire codebase and should ressolve https://github.com/unicode-org/icu4x/issues/1363

gregtatum commented on components/calendar/src/iso.rs:
```diff
@@ -395,6 +395,7 @@ impl Iso {
     }
 
     fn fixed_from_iso_integers(year: i32, month: i32, day: i32) -> i32 {
+        #[allow(clippy::unwrap_used)] // TODO(#1688) Clippy exceptions need docs or fixing.
```
issue: I think this issue number #1688 is incorrect.

gregtatum commented on utils/tinystr/src/ascii.rs:
```diff
@@ -41,6 +41,8 @@ impl<const N: usize> TinyAsciiStr<N> {
         let mut out = [0; N];
         let mut i = 0;
         let mut found_null = false;
+        // Indexing is protected by TinyStrError::TooLarge
```
issue: This is work from #1677.

suggestion: This is already merged into main, and you should be able to merge in the latest, per Shane's suggestion.

gregtatum reviewed (changes requested):
Thanks, this is getting really close. There are still 2 issues that need resolving:

* Updating the issue number to point to the correct follow-up issue.
* Fixing the merge.

It would be nice to get feedback from @sffc that the results of the clippy lints are correct, and the suppressions are used as expected.

ozghimire commented:
@gregtatum The requested changes have been completed and pushed up.

sffc commented on ffi/diplomat/src/lib.rs:
```diff
@@ -7,10 +7,16 @@
+        clippy::upper_case_acronyms
+    )
 )]
-#![no_std]
```
Issue: This is not correct. The `feature(alloc_error_handler)` should be enabled in some cases, but `#![no_std]` and `#![allow(clippy:upper_case_acronyms)]` should always be enabled. Please revert this back to what it was before and add the new clippy exclusions as a standalone cfg_attr.

sffc replied:
Issue: This is still not correct. You should have

```
#![no_std]
#![allow(clippy::upper_case_acronyms)]
#![cfg_attr(
    not(test),
    deny(
        clippy::indexing_slicing,
        clippy::unwrap_used,
        clippy::expect_used,
        clippy::panic
    )
)]
```

sffc commented on provider/cldr/src/lib.rs:
```diff
@@ -19,6 +19,16 @@
 
+#![cfg_attr(
+    not(test),
+    deny(
```
Request: For now, exclude this particular crate (`icu_provider_cldr`) from the change. This crate is being tracked separately in #574

sffc replied:
Request: Please also undo the rest of the changes to the files in this crate.

sffc replied:
Actually, never mind. What you did is fine.

sffc commented on components/calendar/src/iso.rs:
```diff
@@ -395,6 +395,7 @@ impl Iso {
     }
 
     fn fixed_from_iso_integers(year: i32, month: i32, day: i32) -> i32 {
+        #[allow(clippy::unwrap_used)] // TODO(#1668) Clippy exceptions need docs or fixing.
```
Praise: I like having an issue number `#1668` because we can close the issue once all of the TODOs in source code are gone.

sffc reviewed (commented):
Thank you for all your work! Two more comments and then I think we can merge.

ozghimire commented:
@sffc @gregtatum  The changes requested are pushed up and conflicts have been resolved.

sffc reviewed (commented):
Please fix up the diplomat lib.rs and then we can merge; thanks!

sffc reviewed (approved):
LGTM pending CI coming back green

gregtatum commented:
Looks like there are still a few CI issues that need fixing. `cargo fmt` will format the code.

sffc commented:
> @gregtatum @sffc Can anyone approve the CI pipeline run ?

Yes, approved. Sorry for the delay here; once you land your first PR then they should run automatically.

ozghimire commented:
@sffc  These changes should get the CI process to green. Let's run them again.

sffc commented:
There is a conflict; can you resolve the conflict first?

ozghimire commented:
@sffc  @gregtatum  Issue #1363 should be resolved by PR after merge.
